rust/agama-cli/src/logs.rs

```diff
@@ -0,0 +1,257 @@
+    }
+}
+
+const DEFAULT_COMMANDS: [&str; 3] = [
+    "journalctl -u agama",
+    "journalctl -u agama-auto",
+    "journalctl --dmesg",
+];
+
+const DEFAULT_PATHS: [&str; 14] = [
+    // logs
+    "/var/log/YaST2",
+    "/var/log/zypper.log",
+    "/var/log/zypper/history*",
+    "/var/log/zypper/pk_backend_zypp",
+    "/var/log/pbl.log",
+    "/var/log/linuxrc.log",
+    "/var/log/wickedd.log",
+    "/var/log/NetworkManager",
+    "/var/log/messages",
+    "/var/log/boot.msg",
+    "/var/log/udev.log",
+    // config
+    "/etc/install.inf",
+    "/etc/os-release",
+    "/linuxrc.config",
+];
+
+const DEFAULT_RESULT: &str = "/tmp/agama_logs";
+const DEFAULT_COMPRESSION: (&str, &str) = ("bzip2", "tar.bz2");
+const DEFAULT_TMP_DIR: &str = "agama-logs";
+
+// A wrapper around println which shows (or not) the text depending on the boolean variable
+fn showln(show: bool, text: &str) {
+    if !show {
+        return;
+    }
+
+    println!("{}", text);
+}
+
+// A wrapper around println which shows (or not) the text depending on the boolean variable
+fn show(show: bool, text: &str) {
+    if !show {
+        return;
+    }
+
+    print!("{}", text);
+}
+
+// Struct for log represented by a file
+struct LogPath {
+    // log source
+    src_path: String,
+
+    // directory where to collect logs
+    dst_path: PathBuf,
+}
+
+impl LogPath {
+    fn new(src: &str, dst: &Path) -> Self {
+        Self {
+            src_path: src.to_string(),
+            dst_path: dst.to_owned(),
+        }
+    }
+}
+
+// Struct for log created on demmand by a command
+struct LogCmd {
+    // command which stdout / stderr is logged
+    cmd: String,
+
+    // place where to collect logs
+    dst_path: PathBuf,
+}
+
+impl LogCmd {
+    fn new(cmd: &str, dst: &Path) -> Self {
+        Self {
+            cmd: cmd.to_string(),
+            dst_path: dst.to_owned(),
+        }
+    }
+}
+
+trait LogItem {
+    // definition of log source
+    fn from(&self) -> &String;
+
+    // definition of destination as path to a file
+    fn to(&self) -> PathBuf;
+
+    // performs whatever is needed to store logs from "from" at "to" path
+    fn store(&self) -> Result<(), Error>;
+}
+
+impl LogItem for LogPath {
+    fn from(&self) -> &String {
+        &self.src_path
+    }
+
+    fn to(&self) -> PathBuf {
+        // remove leading '/' if any from the path (reason see later)
+        let r_path = Path::new(self.src_path.as_str()).strip_prefix("/").unwrap();
+
+        // here is the reason, join overwrites the content if the joined path is absolute
+        self.dst_path.join(r_path)
+    }
+
+    fn store(&self) -> Result<(), Error> {
+        // for now keep directory structure close to the original
+        // e.g. what was in /etc will be in /<tmp dir>/etc/
+        fs::create_dir_all(self.to().parent().unwrap())?;
+
+        let options = CopyOptions::new();
+        // fs_extra's own Error doesn't implement From trait so ? operator is unusable
```

Comment:
`fs_extra` error does not implement the conversion to `io::Error`. However, implements the [Error trait](https://doc.rust-lang.org/std/error/trait.Error.html), so you can do something like this:

```rust
    fn store(&self) -> Result<(), Box<dyn std::error::Error>> {
        // for now keep directory structure close to the original
        // e.g. what was in /etc will be in /<tmp dir>/etc/
        fs::create_dir_all(self.to().parent().unwrap())?;

        let options = CopyOptions::new();
        copy_items(
            &[self.src_path.as_str()],
            self.to().parent().unwrap(),
            &options,
        )?;
        Ok(())
    }
```
Of course, you can import the error trait as `use std::error::Error` if you wish and you need to adapt signature of the `store` function in the `LogItem` trait.

Another option might be to have your own error type and write the conversion from `fs_extra` error type. You can use [thiserror](https://crates.io/crates/thiserror) to simplify the process.

Reply:
well if written this way it more less masks what is returned by copy_items, doesn't it?

So, I've tried to do the conversion simply by the match rule.

Anyway when ```?``` operator is used on ```copy_items``` it complains "trait From<fs_extra::error::Error>  is not implemented for std::io::Error" so this snippet doesn't work.

Reply:
Well, it does not work because you need to update the `store` signature in the `LogItem` trait. After that, it will work.

About masking the error by `copy_items`, you can have your own error type that keeps the original error if you wish. Actually, [anyhow](https://crates.io/crates/anyhow) allows you to keep the errors backtrace easily.

About error handling, you might be interested in [this article](https://www.lpalmieri.com/posts/error-handling-rust/) by [Luca Palmieri](https://www.lpalmieri.com/). Another link for your reading list 😄.